src/Microsoft.OData.Client/Materialization/MaterializerDeltaResourceSet.cs

```diff
@@ -0,0 +1,97 @@
+using System.Linq;
+
+namespace Microsoft.OData.Client.Materialization
+{
+    /// <summary>
+    /// Materializer state for a given <see cref="ODataDeltaResourceSet"/>.
+    /// </summary>
+    internal class MaterializerDeltaResourceSet : IMaterializerState
+    {
+        /// <summary>The OData delta resource set.</summary>
+        private readonly ODataDeltaResourceSet deltaResourceSet;
+
+        /// <summary>The entries.</summary>
+        private readonly List<IMaterializerState> entries;
+
+        /// <summary>
+        /// Prevents a default instance of the <see cref="MaterializerDeltaResourceSet"/> struct from being created.
+        /// </summary>
+        /// <param name="deltaResourceSet">The OData delta resource set.</param>
+        /// <param name="entries">The entries.</param>
+        private MaterializerDeltaResourceSet(ODataDeltaResourceSet deltaResourceSet, List<IMaterializerState> entries)
+        {
+            Debug.Assert(deltaResourceSet != null, "deltaResourceSet != null");
+            Debug.Assert(entries != null, "entries != null");
+
+            this.deltaResourceSet = deltaResourceSet;
+            this.entries = entries;
+        }
+
+        /// <summary>
+        /// Gets the delta resource set.
+        /// </summary>
+        public ODataDeltaResourceSet DeltaResourceSet
+        {
+            get { return this.deltaResourceSet; }
+        }
+
+        /// <summary>
+        /// Gets the entries.
+        /// </summary>
+        public List<IMaterializerState> Entries
+        {
+            get { return this.entries; }
+        }
+
+        /// <summary>
+        /// Adds a <see cref="IMaterializerState"/> entry to the entries collection.
+        /// </summary>
+        /// <param name="entry">The <see cref="IMaterializerState"/> entry to be added.</param>
+        public void AddEntry(IMaterializerState entry)
+        {
+            this.entries.Add(entry);
+        }
+
+        /// <summary>
+        /// Creates the delta resource set.
+        /// </summary>
+        /// <param name="deltaResourceSet">The OData delta resource set to be created.</param>
+        /// <param name="entries">The entries of an <see cref="ODataDeltaResourceSet"/>.</param>
+        /// <param name="materializerContext">The current materializer context.</param>
+        /// <returns>The <see cref="MaterializerDeltaResourceSet"/> of the created <see cref="ODataDeltaResourceSet"/>.</returns>
+        public static MaterializerDeltaResourceSet CreateODataDeltaResourceSet(ODataDeltaResourceSet deltaResourceSet, List<IMaterializerState> entries, IODataMaterializerContext materializerContext)
+        {
+            Debug.Assert(materializerContext.GetAnnotation<HashSet<IMaterializerState>>(deltaResourceSet) == null, "DeltaResourceSet state has already been created.");
+            if (entries == null)
+            {
+                entries = (List<IMaterializerState>)Enumerable.Empty<IMaterializerState>();

```

Comment:
I don't think it's safe to cast `Enumerable.Empty<T>` to a `List<T>`. This is probably going to throw an `InvalidCastException` at runtime. Have you tested this code path?